### PR TITLE
Making tasks to show up in oldest first manner, highlighting overdue tasks.

### LINKED ISSUE #155: Arrange the issues in order of the most time elapsed since issue was opened.
The issues that have the most amount of time elapsed since opening should have a slightly red background and be on tope of the list, followed by newer tasks.

### PR DESCRIPTION
**Pull request by Devashish0210**
Link for the issue : [github.com/Real-Dev-Squad/website-status/issues/155](url)

Closes #155 

A screenshot of the status site before my feature has been implemented : 
image

A screenshot of the local status site after my changes have been implemented : 
image

I have arranged the tasks in oldest first manner, and highlighted the overdue tasks.

### CODE REVIEW

File: src/pages/index.tsx [hunk 1]
```diff
@@ -6,6 +6,7 @@ import useFetch from '@/hooks/useFetch';
 import classNames from '@/styles/tasks.module.scss';
 import task from '@/interfaces/task.type';
 import Accordion from '@/components/Accordion';
+// import { COMPLETED } from '@/components/constants/task-status'; unused variable.
```
Comment (harshith-venkatesh) [inline src/pages/index.tsx:9]: we can remove this comment if we are not using it

File: src/components/tasks/card/card.module.scss [hunk 2]
```diff
@@ -6,9 +6,10 @@
   cursor: pointer;
   margin-bottom: 25px;
   justify-content: space-between;
+  border: solid 2px;
   border-bottom-style: solid;
   border-bottom-color: #d9d9d9;
-  border-bottom-width: 1px;
+  border-color: whitesmoke;
```
Comment (harshith-venkatesh) [inline src/components/tasks/card/card.module.scss:12]: Learnt about new border-color name.
Would it be better to go ahead with #f5f5f5

File: src/components/tasks/card/index.tsx [hunk 3]
```diff
@@ -33,18 +31,30 @@ const Card: FC<Props> = ({ content }) => {
...
+  const cardClassNames = [classNames.card];
+  const currentDate = new Date();
+  if (status !== COMPLETED) {
+    if (+localEndsOn - (+currentDate) <= 0) {
+      cardClassNames.push(classNames.overdueTask);
+    }
+  }
   return (
-    <div className={classNames.card}>
+    <div className={cardClassNames.join(' ')} id="wrapper-div">
```
Comment (harshith-venkatesh) [inline src/components/tasks/card/index.tsx:49]: What is the purpose of adding id here?

File: src/components/tasks/card/card.module.scss [hunk 4]
```diff
@@ -52,3 +52,8 @@
   font-size: 1.3rem;
   font-weight: 500;
 }
+
+.overdueTask{
+  border: solid red 2px;
+   background-color: #ffebee;
```
Comment (whyDontI) [inline src/components/tasks/card/card.module.scss:58]: Can you please fix the indentation?

File: src/components/tasks/card/index.tsx [hunk 5]
```diff
@@ -12,19 +12,17 @@ import {
...
-    endsOn,
-    startedOn,
-    status,
-    assignee,
+    title, endsOn, startedOn, status, assignee,
   } = content;
 
-  const [assigneeProfilePic, setAssigneeProfilePic] = useState(`${process.env.NEXT_PUBLIC_GITHUB_IMAGE_URL}${assignee}/img.png`);
+  const [assigneeProfilePic, setAssigneeProfilePic] = useState(
+    `${process.env.NEXT_PUBLIC_GITHUB_IMAGE_URL}${assignee}/img.png`,
```
Comment (whyDontI) [inline src/components/tasks/card/index.tsx:24]: While you are at it, can you remove the last `/` from the `NEXT_PUBLIC_GITHUB_IMAGE_URL` variable?

Files -
https://github.com/Real-Dev-Squad/website-status/blob/develop/.env.development
https://github.com/Real-Dev-Squad/website-status/blob/develop/.env.production

File: src/components/tasks/card/index.tsx [hunk 6]
```diff
@@ -57,17 +67,17 @@ const Card: FC<Props> = ({ content }) => {
...
         </span>
       </div>
       <div className={classNames.cardItems}>
         <span className={classNames.cardSpecialFont}>
           Started
           {' '}
-          {fromNowStartedOn}
+          {
+            fromNowStartedOn
+          }
```
Comment (whyDontI) [inline src/components/tasks/card/index.tsx:80]: Can we revert this? not a required change I guess.

File: src/pages/index.tsx [hunk 7]
```diff
@@ -22,6 +23,7 @@ const Index: FC = () => {
   useEffect(() => {
     if ('tasks' in response) {
       tasks = response.tasks;
+      tasks.sort((a, b) => (+a.endsOn - (+b.endsOn)));
```
Comment (harshith-venkatesh) [inline src/pages/index.tsx:26]: We need specify the type of params for a and b

File: src/pages/index.tsx [hunk 8]
```diff
@@ -40,23 +42,19 @@ const Index: FC = () => {
...
+          <>
+            {Object.keys(filteredTask).length > 0
+              ? Object.keys(filteredTask).map((key) => (
+                <Accordion open title={key} key={key}>
+                  {renderCardList(filteredTask[key])}
+                </Accordion>
+              ))
+              : !error && 'No Tasks Found'}
+          </>
+        )}
```
Comment (whyDontI) [inline src/pages/index.tsx:57]: Can we revert these indentation changes?

File: src/pages/index.tsx [hunk 9]
```diff
@@ -22,6 +23,7 @@ const Index: FC = () => {
   useEffect(() => {
     if ('tasks' in response) {
       tasks = response.tasks;
+      tasks.sort((a, b) => (+a.endsOn - (+b.endsOn)));
       const taskMap: any = [];
```
Comment (whyDontI) [inline src/pages/index.tsx:27]: While you are at it, can you add a proper type for the `taskMap`?
https://www.typescriptlang.org/docs/handbook/basic-types.html

File: src/components/tasks/card/index.tsx [hunk 10]
```diff
@@ -57,17 +67,17 @@ const Card: FC<Props> = ({ content }) => {
             width={iconWidth}
             height={iconHeight}
           />
-          <span className={classNames.cardSpecialFont}>
-            Due Date
-          </span>
+          <span className={classNames.cardSpecialFont}>Due Date</span>
```
Comment (whyDontI) [inline src/components/tasks/card/index.tsx:70]: Can we revert the indentation change?

File: src/components/tasks/card/index.tsx [hunk 11]
```diff
@@ -33,18 +31,30 @@ const Card: FC<Props> = ({ content }) => {
...
+    <div className={cardClassNames.join(' ')} id="wrapper-div">
       <div className={classNames.cardItems}>
         <span className={classNames.cardTitle}>{title}</span>
         <span>
           <span className={classNames.cardSpecialFont}>Status:</span>
-          <span className={classNames.cardStatusFont} style={{ color: statusFontColor }}>
+          <span
+            className={classNames.cardStatusFont}
+            style={{ color: statusFontColor }}
+          >
```
Comment (whyDontI) [inline src/components/tasks/card/index.tsx:57]: Can we revert the indentation change?

File: src/components/tasks/card/index.tsx [hunk 12]
```diff
@@ -33,18 +31,30 @@ const Card: FC<Props> = ({ content }) => {
...
+    ? '#00a337'
+    : '#f83535';
 
   const iconHeight = '25px';
   const iconWidth = '25px';
 
+  const cardClassNames = [classNames.card];
+  const currentDate = new Date();
+  if (status !== COMPLETED) {
+    if (+localEndsOn - (+currentDate) <= 0) {
```
Comment (whyDontI) [inline src/components/tasks/card/index.tsx:44]: Is there any special reason to use `+` before the variables?

File: src/components/tasks/card/index.tsx (hunk 5 above)
Comment (whyDontI) [reply src/components/tasks/card/index.tsx:24]: Also please revert the indentation change if not necessary

**Review by whyDontI — CHANGES_REQUESTED** [review]
Good work 👍🏽 
Requesting some changes

File: src/pages/index.tsx (hunk 1 above)
Comment (Devashish0210) [reply src/pages/index.tsx:9]: For sure yes. I kept that there in case we needed it in the future 😅

File: src/components/tasks/card/card.module.scss (hunk 2 above)
Comment (Devashish0210) [reply src/components/tasks/card/card.module.scss]: Looked good to me 😅 I'll do some more CSS tasks and get comfortable using codes rather than the names

File: src/components/tasks/card/index.tsx (hunk 3 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:49]: to get the element by id. I had a previous logic which needed the id. It is not needed now. I will remove it. Thanks 😁

File: src/components/tasks/card/card.module.scss (hunk 4 above)
Comment (Devashish0210) [reply src/components/tasks/card/card.module.scss:58]: On it. Thanks

File: src/components/tasks/card/index.tsx (hunk 6 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:80]: it showed me a red squiggly line saying that we can only have one jsx expression per line. I'll change it

File: src/components/tasks/card/index.tsx (hunk 12 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:44]: I'll test it without the + symbol and let you know 😁

File: src/components/tasks/card/index.tsx (hunk 11 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:57]: For sure 😁 I'll disable prettier for the entire file and revert all the changes

File: src/pages/index.tsx (hunk 8 above)
Comment (Devashish0210) [reply src/pages/index.tsx:57]: Made changes. Please check 😁

File: src/pages/index.tsx (hunk 1 above)
Comment (harshith-venkatesh) [reply src/pages/index.tsx:9]: @Devashish0210 , you can resolve conversation if you have solved this

File: src/components/tasks/card/index.tsx (hunk 10 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:70]: Reverted. Please check.

File: src/components/tasks/card/index.tsx (hunk 5 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:24]: indentation done. /'s removed. Please check.

File: src/components/tasks/card/index.tsx (hunk 6 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx]: reverted. Please check.

File: src/components/tasks/card/index.tsx (hunk 12 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx]: refactored. Please check

File: src/pages/index.tsx (hunk 7 above)
Comment (Devashish0210) [reply src/pages/index.tsx]: Specified the types of both a and b. 😁

File: src/pages/index.tsx (hunk 9 above)
Comment (Devashish0210) [reply src/pages/index.tsx:27]: As we had discussed about this issue on the call, I will mark this as resolved.

File: src/components/tasks/card/card.module.scss (hunk 4 above)
Comment (Devashish0210) [reply src/components/tasks/card/card.module.scss]: Fixed 😁

File: src/components/tasks/card/index.tsx (hunk 11 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:57]: reverted. please check.

File: src/components/tasks/card/index.tsx [hunk 13]
```diff
@@ -38,8 +34,15 @@ const Card: FC<Props> = ({ content }) => {
   const iconHeight = '25px';
   const iconWidth = '25px';
 
+  const cardClassNames = [classNames.card];
+  const currentDate = new Date();
+  if (status !== COMPLETED) {
+    if (localEndsOn.valueOf() - currentDate.valueOf() <= 0) {
+      cardClassNames.push(classNames.overdueTask);
+    }
+  }
```
Comment (whyDontI) [inline src/components/tasks/card/index.tsx:43]: Move this logic to a separate function called `isTaskOverdue(status, localEndsOn)` then use this function on line number 45 to add class.

File: src/components/tasks/card/index.tsx [hunk 14]
```diff
@@ -38,8 +34,15 @@ const Card: FC<Props> = ({ content }) => {
...
   return (
-    <div className={classNames.card}>
+    <div className={cardClassNames.join(' ')}>
```
Comment (whyDontI) [inline src/components/tasks/card/index.tsx:45]: Once you define `isTaskOverdue` function, use it here for conditional class names using template literals.

```HTML
<div className={`
    ${classNames.card}
    ${isTaskOverdue() && classNames.overdueTask}
`}>
.
.
</div>
```

We'll get someone to write a utility function for this kind of case in the future.

File: src/components/tasks/card/index.tsx [hunk 15]
```diff
@@ -66,7 +69,6 @@ const Card: FC<Props> = ({ content }) => {
       <div className={classNames.cardItems}>
         <span className={classNames.cardSpecialFont}>
           Started
-          {' '}
```
Comment (whyDontI) [inline src/components/tasks/card/index.tsx:69]: As you removed this space, it would look like this
`Started3 months ago`
Can you please add the space again?

File: src/pages/index.tsx (hunk 8 above)
Comment (whyDontI) [reply src/pages/index.tsx:56]: Can't see any changes

File: .env.development [hunk 16]
```diff
@@ -1,2 +1,2 @@
-NEXT_PUBLIC_BASE_URL='https://staging-api.realdevsquad.com'
-NEXT_PUBLIC_GITHUB_IMAGE_URL='https://raw.githubusercontent.com/Real-Dev-Squad/website-static/main/members/'
+NEXT_PUBLIC_BASE_URL='https://api.realdevsquad.com'
```
Comment (whyDontI) [inline .env.development:1]: We use `staging-api` in development environment.
Can you please revert this change back?

**Review by whyDontI — CHANGES_REQUESTED** [review]
Requested some changes

File: src/components/tasks/card/index.tsx (hunk 15 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:69]: Added.

File: .env.development (hunk 16 above)
Comment (Devashish0210) [reply .env.development:1]: Resolved.

File: src/components/tasks/card/index.tsx (hunk 13 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:43]: Made the changes. Please check

Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:43]: Made the changes, please check

File: src/components/tasks/card/index.tsx (hunk 14 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:45]: Done 😁

File: src/components/tasks/card/index.tsx [hunk 17]
```diff
@@ -38,8 +34,23 @@ const Card: FC<Props> = ({ content }) => {
   const iconHeight = '25px';
   const iconWidth = '25px';
 
+  function isTaskOverdue() {
+    const cardClassNames = [classNames.card];
+    const currentDate = new Date();
+    if (status !== COMPLETED) {
+      if (localEndsOn.valueOf() - currentDate.valueOf() <= 0) {
+        cardClassNames.push(classNames.overdueTask);
+        return true;
```
Comment (ankushdharkar) [inline src/components/tasks/card/index.tsx:43]: `cardClassNames` seems to be a variable local to the `isTaskOverdue()`. What do we gain by setting it?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: src/components/tasks/card/index.tsx (hunk 17 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:43]: cardClassNames adds the "overdue-task" class name to the already existing class name of the overdue tasks div tag. It's an array that already holds the current class name of the div tag and if overdue, joins the overdue-task class name. Then the css for overdue task is added to it.

Comment (ankushdharkar) [reply src/components/tasks/card/index.tsx:43]: `isTaskOverdue()` is a getter function. It should only return a value and not perform any [side-effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science))

Based on what the return value is, we should manipulate the array you pushed to, not while getting the boolean state

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: src/components/tasks/card/index.tsx (hunk 17 above)
Comment (Devashish0210) [reply src/components/tasks/card/index.tsx]: Have made the necessary changes. Requesting re-evaluation 😁

File: src/components/tasks/card/index.tsx [hunk 18]
```diff
@@ -38,8 +34,26 @@ const Card: FC<Props> = ({ content }) => {
...
+  const cardClassNames = [classNames.card];
+
+  function isTaskOverdue() {
+    const currentDate = new Date();
+    if (status !== COMPLETED) {
+      if (localEndsOn.valueOf() - currentDate.valueOf() <= 0) {
+        return true;
+      }
+    }
+    return false;
```
Comment (ankushdharkar) [inline src/components/tasks/card/index.tsx:46]: Nit: You can reduce it to

```js
 function isTaskOverdue() {
    const currentDate = new Date();
    const timeLeft = localEndsOn.valueOf() - currentDate.valueOf();
    return status !== COMPLETED && timeLeft <= 0;
}
```

Comment (Devashish0210) [reply src/components/tasks/card/index.tsx:46]: Awesome. Making the requested changes.
